clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -554,7 +664,13 @@ private static int parseAcks(String acksString) {
             long timestamp = record.timestamp() == null ? time.milliseconds() : record.timestamp();
             log.trace("Sending record {} with callback {} to topic {} partition {}", record, callback, record.topic(), partition);
             // producer callback will make sure to call both 'callback' and interceptor callback
-            Callback interceptCallback = this.interceptors == null ? callback : new InterceptorCallback<>(callback, this.interceptors, tp);
+            Callback interceptCallback = new InterceptorCallback<>(callback, this.interceptors, tp, transactionManager);
+
+            if (transactionManager != null && transactionManager.isInTransaction()) {
+                transactionManager.maybeAddPartitionToTransaction(new TopicPartition(record.topic(), record.partition()));
```

Comment:
This breaks with an NPE when the user does not specify the partition explicitly in the `ProducerRecord`. We should use `tp` as the partition.